Add a Changelog

Finnerale opened the pull request:
This introduces an in-tree changelog base on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

Whenever a new PR ist created it should contain changelog entries for everything it changes, thus keeping the changelog always in sync with `master`.

Every changelog entry is annotated with the PR that introduced it and the author for attribution.

The attribution is also the reason why the `Maintenance` section exists, even though entries in it have no *direct* impact on users.

Annotating entries with a link to the introducing PR causes a chicken-and-egg problem tho.
When creating the PR you wont know the number yet, which you need to refer to it.
Having a quick link to the related PR seems very useful however as it provides much more information about the change than the changelog ever could.

See [Zulip](https://xi.zulipchat.com/#narrow/stream/147926-druid/topic/Bookkeeping.20about.20breaking.20changes) for previous discussion.

xStrom commented:
The chicken and egg problem isn't that bad because github issue numbers are sequential and you can guess what number you will get. Sometimes you lose the race, but that should be rare.

Finnerale commented:
Right, they are sequential that makes it a bit better but then you would have to figure out what the current count is at, I wonder if there is like a 'next PR number' badge that we could add to the readme for example.

xStrom commented on CHANGELOG.md:
```diff
@@ -0,0 +1,162 @@
+# Changelog
+
+All notable changes to this project will be documented in this file.
+
+The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
+and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
```
Do we really want this preface text at the top? Other notable projects like cargo don't have anything like that in [their changelog](https://github.com/rust-lang/cargo/blob/master/CHANGELOG.md).

If anything I think an adaptation of this section could live in the main readme. Perhaps something like the following at the bottom of the first introduction section:
```markdown
Druid adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html) and documents
all its changes in [the changelog](https://github.com/xi-editor/druid/blob/master/CHANGELOG.md).
```

Finnerale replied:
I've removed the preface.
Linking the changelog seams like a good idea.
Considering that semantic versioning is standard for Rust anyways I don't think that has to be mentioned.

Finnerale replied:
I've now also added a link to the readme.

xStrom commented on CHANGELOG.md:
```diff
@@ -0,0 +1,162 @@
+
+#### Mostly complete Wasm backend for druid-shell. ([#759])
+
+[@elrnv] continued the work of [@tedsta] and implemented a mostly complete Wasm backend and enabled all druid example to [run in the browser](https://elrnv.github.io/druid-wasm-examples/).
```
~example~ examples

xStrom commented on CHANGELOG.md:
```diff
@@ -0,0 +1,162 @@
+### Added
+
+- `TextBox` can receive `EditAction` commands. ([#814] by [@cmyr])
+- `Split::min_splitter_area(f64)` to add padding around the split bar. ([#738] by [@xStorm])
```
~split bar~ splitter bar

xStrom commented on CHANGELOG.md:
```diff
@@ -0,0 +1,162 @@
+- `TextBox` can receive `EditAction` commands. ([#814] by [@cmyr])
+- `Split::min_splitter_area(f64)` to add padding around the split bar. ([#738] by [@xStorm])
+- Published `druid::text` module. ([#816] by [@cmyr])
+- Basic X11 backend for druid-shell. ([#599] by [@crsaracco])
```
Maybe already highlighted changes don't need to show up in the listing?

xStrom commented on CHANGELOG.md:
```diff
@@ -0,0 +1,162 @@
+- `Split::min_splitter_area(f64)` to add padding around the split bar. ([#738] by [@xStorm])
+- Published `druid::text` module. ([#816] by [@cmyr])
+- Basic X11 backend for druid-shell. ([#599] by [@crsaracco])
+- `InternalEvent::MouseLeave` signalling the cursor leaved the window. ([#821] by [@teddemunnik])
```
~leaved~ left

xStrom commented on CHANGELOG.md:
```diff
@@ -0,0 +1,162 @@
+- Basic X11 backend for druid-shell. ([#599] by [@crsaracco])
+- `InternalEvent::MouseLeave` signalling the cursor leaved the window. ([#821] by [@teddemunnik])
+- `children_changed` now always includes layout and paint request. ([#839] by [@xStorm])
+- Mostly complete Wasm backend for druid-shell. ([#759] by [@elrnv])
```
Maybe already highlighted changes don't need to show up in the listing?

xStrom commented on CHANGELOG.md:
```diff
@@ -0,0 +1,162 @@
+- Start focus cycling from non-registered-for-focus widgets. ([#819] by [@xStorm])
+- Propagate `Event::FocusChanged` to focus gaining widgets as well. ([#819] by [@xStorm])
+- GTK: Prevent crashing on pop-ups. ([#837] by [@finnerale])
+- Keep hot state  consistent with mouse position. ([#841] by [@xStorm])
```
Remove double space.

xStrom commented on CHANGELOG.md:
```diff
@@ -0,0 +1,162 @@
+- Keep hot state  consistent with mouse position. ([#841] by [@xStorm])
+- Open file menu item works again. ([#851] by [@kindlychung])
+- Supply correct `LifeCycleCtx` to `Event::FocusChanged`. ([#878] by [@cmyr])
+- Windows: Termiate app when all windows have closed. ([#763] by [@xStorm])
```
This same #763 ended up containing a macOS fix too. Perhaps add the following line:
```markdown
- macOS: `Application::quit` now quits the run loop instead of killing the process. ([#763] by [@xStrom])
```

xStrom commented on CHANGELOG.md:
```diff
@@ -0,0 +1,162 @@
+- Window title can be any `LabelText` (such as a simple `String`). ([#869] by [@cmyr])
+- `Label::with_font` and `set_font`. ([#785] by [@thecodewarrior])
+
+### Changed
```
There's another significant change:

```markdown
- Global `Application` associated functions are now instance methods instead, e.g. `Application::global().quit()` instead of the old `Application::quit()`. ([#763] by [@xStrom])
```

xStrom commented on CHANGELOG.md:
```diff
@@ -0,0 +1,162 @@
+### Visual
+
+- Improved `Split` accuracy. ([#738] by [@xStorm])
+- Build-in widgets no longer stroke outside their `paint_rect`. ([#861] by [@jneem])
```
~Build-in~ Built-in

xStrom commented on CHANGELOG.md:
```diff
@@ -0,0 +1,162 @@
+- Added rendering tests. ([#784] by [@fishrockz])
+- CI testing has been revamped. ([#857] by [@xStorm])
+- Associate timers with widget ids. ([#831] by [@sjoshid])
+- Add a changelog. ([#889] by [@finnerale])
```
I don't think future tense makes sense for a changelog. So this should be *Added* not *Add*. Should probably check other lines too. Generally speaking I guess both past tense (*Added a changelog*) and present tense (*There is now a changelog*) are fine and both can be used, but yeah not future.

Finnerale replied:
Yea I thing I've mixed tenses a bit :sweat_smile: 
I'm not quite sure what fits better (present or past tense).

Finnerale commented:
@xStrom I've fixed your name (sorry for that), added some other things I've missed and hopefully corrected the tenses.

I'm not sure if we should include changes to internal-but-public API such as `InternalEvent`. For now I've just added them.

xStrom commented:
We should also add a section/note to `CONTRIBUTING.md` about needing to change the changelog in PRs.

I think documenting the internal stuff is fine while we're in the 0.x versions. We can reconsider this later, e.g. once we hit 1.0.

Also #890 got merged.

Finnerale commented:
I've added it to `CONTRIBUTING.md`.

I've not added anything regarding #890 as it is a follow-up for a feature that is not released yet.
I think keeping the changelog compact and only freezing entries once the version is released makes it easier to read. For example I've merged the `Event::MouseLeave` addition with your effort to collect `InternalEvent`s. Thus there is only one entry mentioning the addition of `InternalEvent::MouseLeave`.

xStrom commented:
I think that makes sense. We don't want the changelog to be a mirror copy of the merge commit log. Multiple PRs working on the same conceptual feature can be streamlined into a single entry in the changelog.

xStrom commented on CHANGELOG.md:
```diff
@@ -0,0 +1,159 @@
+[#878]: https://github.com/xi-editor/druid/pull/878
+[#889]: https://github.com/xi-editor/druid/pull/899
+
+## [0.5] - 2020-04-01
```
I think we should always use the full semver version. `0.5` -> `0.5.0`.

xStrom commented:
For completeness sake, let's add [all the previous versions](https://crates.io/crates/druid/versions) to the changelog as well. I know we don't have the the actual changes for them right now, but we could at least have the version numbers and their dates.

xStrom commented on CHANGELOG.md:
```diff
@@ -134,10 +134,17 @@ While some features like the clipboard, menus or file dialogs are not yet availa
 
+## [0.4.0] - 2019-12-28
+## [0.3.2] - 2019-11-05
+## [0.3.1] - 2019-11-04
```
Shouldn't this be 2019-11-05 as well? :thinking:

Finnerale replied:
@xStrom I'm not sure what you mean, it seems correct to me.

xStrom replied:
Haha, well if it's not a typo then both GitHub and crates.io show the date in my timezone. That's why I see 2019-11-05.
